dart/lib/src/scope.dart

```diff
@@ -0,0 +1,210 @@
+import 'dart:collection';
+
+import 'protocol.dart';
+import 'sentry_options.dart';
+
+/// Scope data to be sent with the event
+class Scope {
+  /// How important this event is.
+  SentryLevel level;
+
+  /// The name of the transaction which generated this event,
+  /// for example, the route name: `"/users/<username>/"`.
+  String transaction;
+
+  /// Information about the current user.
+  User user;
+
+  /// Used to deduplicate events by grouping ones with the same fingerprint
+  /// together.
+  ///
+  /// Example:
+  ///
+  ///     // A completely custom fingerprint:
+  ///     var custom = ['foo', 'bar', 'baz'];
+  List<String> _fingerprint;
+
+  List<String> get fingerprint =>
+      _fingerprint != null ? List.unmodifiable(_fingerprint) : null;
+
+  set fingerprint(List<String> fingerprint) {
+    _fingerprint = fingerprint;
+  }
+
+  /// List of breadcrumbs for this scope.
+  ///
+  /// See also:
+  /// * https://docs.sentry.io/enriching-error-data/breadcrumbs/?platform=javascript
+  final Queue<Breadcrumb> _breadcrumbs = Queue();
+
+  /// Unmodifiable List of breadcrumbs
+  List<Breadcrumb> get breadcrumbs => List.unmodifiable(_breadcrumbs);
+
+  /// Name/value pairs that events can be searched by.
+  final Map<String, String> _tags = {};
+
+  Map<String, String> get tags => Map.unmodifiable(_tags);
+
+  /// Arbitrary name/value pairs attached to the scope.
+  ///
+  /// Sentry.io docs do not talk about restrictions on the values, other than
+  /// they must be JSON-serializable.
+  final Map<String, dynamic> _extra = {};
+
+  Map<String, dynamic> get extra => Map.unmodifiable(_extra);
+
+  // TODO: Contexts
+
+  /// Scope's event processor list
+  final List<EventProcessor> _eventProcessors = [];
+
+  List<EventProcessor> get eventProcessors =>
+      List.unmodifiable(_eventProcessors);
+
+  final SentryOptions _options;
+
+  Scope(this._options) : assert(_options != null, 'SentryOptions is required');
+
+  /// Adds a breadcrumb to the breadcrumbs queue
+  void addBreadcrumb(Breadcrumb breadcrumb, {dynamic hint}) {
+    assert(breadcrumb != null, "Breadcrumb can't be null");
+
+    // bail out if maxBreadcrumbs is zero
+    if (_options.maxBreadcrumbs == 0) {
+      return;
+    }
+
+    // run before breadcrumb callback if set
+    if (_options.beforeBreadcrumbCallback != null) {
+      breadcrumb = _options.beforeBreadcrumbCallback(breadcrumb, hint);
```

Comment:
Do we need `Callback` in the name?

```suggestion
      breadcrumb = _options.beforeBreadcrumb(breadcrumb, hint);
```